### PR TITLE
[vulkan] Pop markers before commands flush

### PR DESCRIPTION
**Pull request by poweifeng**
The flush call to command buffer resets the mCurrent pointer in VulkanCommands such that popGroupMarker will actually trigger the creation of a new buffer.

Before we flush (and reset the client-side state), we should pop group markers if there is one.

This needs a follow-up where we need to make sure that the pop after flush should never happen.

### CODE REVIEW

File: RELEASE_NOTES.md [hunk 1]
```diff
@@ -17,6 +17,7 @@ A new header is inserted each time a *tag* is created.
              normalized position and keeps z reversed [⚠️ **Recompile Materials**]
 - backend: workaround Adreno shader compiler bug (#6355) [⚠️ **Recompile Materials**]
 - geometry: change computing tangent basis from normal vector to use Frisvad's method
+- backend: workaround marker/label pop ordering bug (6292)
```
Comment (pixelflinger) [inline RELEASE_NOTES.md:20]:
```suggestion
- backend: workaround marker/label pop ordering bug (#6292)
```

File: filament/backend/src/vulkan/VulkanCommands.cpp [hunk 2]
```diff
@@ -154,6 +154,17 @@ bool VulkanCommands::flush() {
     const int64_t index = mCurrent - &mStorage[0];
     VkSemaphore renderingFinished = mSubmissionSignals[index];
 
+    // VUID-vkEndCommandBuffer-commandBuffer-01815
+    // TODO: We need to order calls so that there is no outstanding label or markers.
+    if (!mCurrentDebugLabel.empty()) {
```
Comment (pixelflinger) [inline filament/backend/src/vulkan/VulkanCommands.cpp:159]:
```suggestion
    if (UTILS_UNLIKELY(!mCurrentDebugLabel.empty())) {
```

File: filament/backend/src/vulkan/VulkanCommands.cpp [hunk 3]
```diff
@@ -154,6 +154,17 @@ bool VulkanCommands::flush() {
...
+      endDebugUtilsLabel();
+    }
+
+    // VUID-vkEndCommandBuffer-commandBuffer-00062
+    if (!mCurrentDebugMarker.empty()) {
```
Comment (pixelflinger) [inline filament/backend/src/vulkan/VulkanCommands.cpp:164]:
```suggestion
    if (UTILS_UNLIKELY(!mCurrentDebugMarker.empty())) {
```

**Review by pixelflinger — APPROVED** [review]
Approved, but I had two small comments.

**Comment by pixelflinger** [comment]
Is this a fix for #6192 or #6292 ??

**Comment by poweifeng** [comment]
#6192 actually.  I'll update the `RELEASE_NOTES.md`

File: filament/backend/src/vulkan/VulkanCommands.cpp (hunk 2 above)
Comment (poweifeng) [reply filament/backend/src/vulkan/VulkanCommands.cpp:159]: looks like `UTILS_UNLIKELY` is a compiler hint for branching, but we're actually going to hit this branch on almost every frame. Still add it?

File: filament/backend/src/vulkan/VulkanCommands.cpp [hunk 4]
```diff
@@ -154,6 +154,17 @@ bool VulkanCommands::flush() {
     const int64_t index = mCurrent - &mStorage[0];
     VkSemaphore renderingFinished = mSubmissionSignals[index];
 
+    // VUID-vkEndCommandBuffer-commandBuffer-01815
+    // TODO: We need to order calls so that there is no outstanding label or markers.
+    if (!mCurrentDebugLabel.empty()) {
+      endDebugUtilsLabel();
```
Comment (bejado) [inline filament/backend/src/vulkan/VulkanCommands.cpp:160]: 4 spaces indent, please

File: filament/backend/src/vulkan/VulkanCommands.cpp [hunk 5]
```diff
@@ -154,6 +154,17 @@ bool VulkanCommands::flush() {
...
+    }
+
+    // VUID-vkEndCommandBuffer-commandBuffer-00062
+    if (!mCurrentDebugMarker.empty()) {
+      endDebugMarker();
```
Comment (bejado) [inline filament/backend/src/vulkan/VulkanCommands.cpp:165]: 4 spaces indent

File: filament/backend/src/vulkan/VulkanCommands.cpp [hunk 6]
```diff
@@ -271,6 +282,93 @@ void VulkanCommands::updateFences() {
...
+void VulkanCommands::endDebugUtilsLabel() {
+    // TODO: the following should be turned into asserts after refactoring the push/pop
+    // label flow.
+    if (!mCurrent) {
+	return;
+    }
+    if (!mCurrent->cmdbuffer) {
+	return;
+    }
+    if (!mCurrentDebugLabel.empty()) {
```
Comment (bejado) [inline filament/backend/src/vulkan/VulkanCommands.cpp:319]: Should this be `if (mCurrentDebugLabel.empty())` ? Same question inside `endDebugMarker`.

File: filament/backend/src/vulkan/VulkanCommands.cpp (hunk 5 above)
Comment (poweifeng) [reply filament/backend/src/vulkan/VulkanCommands.cpp:165]: done

File: filament/backend/src/vulkan/VulkanCommands.cpp (hunk 4 above)
Comment (poweifeng) [reply filament/backend/src/vulkan/VulkanCommands.cpp:160]: done

File: filament/backend/src/vulkan/VulkanCommands.cpp (hunk 2 above)
Comment (pixelflinger) [reply filament/backend/src/vulkan/VulkanCommands.cpp:159]: Well, after understanding better that this is actually more a workaround, I think it doesn't matter. Eventually this code should go away (maybe an assert), once we figure out a better API for debug markers. So I'll leave up to you.

**Review by pixelflinger — APPROVED** [review]

**Comment by poweifeng** [comment]
Abandoning PR to do a different workaround

**Pull request closed without merging**